File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D06.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getInts

class Y21D06(input: String) : Day {

    private val fish: Map<Int, Int> = input
        .getInts()
        .groupingBy { it }
        .eachCount()

    // initialize count of fish by day
    private val fishByDays: List<Long> = List(9) { i -> fish[i]?.toLong() ?: 0L }

    // iterate one day
    private fun List<Long>.propagate() = List(9) { i ->
        when (i) {
            6 -> get(0) + get(7)
            8 -> get(0)
            else -> get(i + 1)
        }
    }

    fun solve(days: Int): Long = generateSequence(fishByDays) { it.propagate() }
        .take(days + 1)
        .last()
        .sum()

    override fun part1() = solve(80)

    override fun part2() = solve(256)
}

fun main() = Day.runDay(Y21D06::class)

//    Class creation: 16ms
//    Part 1: 361169 (0ms)
//    Part 2: 1634946868992 (0ms)
//    Total time: 17ms

File: rust/advent/src/bin/2021/y21d06.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input = [usize; 9];
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 6).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let mut fish = [0; 9];
    for day in input.get_numbers::<usize>() {
        fish[day] += 1; 
    }
    fish
}

fn solve(fish: &[usize; 9], days: usize) -> Output {
    successors(Some(*fish), |fish| {
        Some(propagate(fish))
    })
    .take(days + 1)
    .last()
    .unwrap()
    .iter()
    .sum()
}

fn propagate(fish: &[usize; 9]) -> [usize; 9] {
    let mut next = [0; 9];
    for day in 0..9 {
        next[day] = match day {
            6 => fish[0] + fish[7],
            8 => fish[0],
            d => fish[d + 1],
        };
    }
    next
}

fn part1(input: &Input) -> Output {
    solve(input, 80)
}

fn part2(input: &Input) -> Output {
    solve(input, 256)
}

#[test]
fn default() {
    let input = get_input(21, 6).unwrap();
    let input = parse_input(&input);
    assert_eq!(361169, part1(&input));
    assert_eq!(1634946868992, part2(&input));
}

// Input parsed (15μs)
// 1. 361169 (5μs)
// 2. 1634946868992 (1μs)
// Total: 24μs